homeassistant/components/pglab/discovery.py

```diff
@@ -0,0 +1,277 @@
+        """Destroy the device and any enties connected with the device."""
+
+        if device_id not in self._discovered:
+            return
+
+        discovery_info = self._discovered[device_id]
+
+        # destroy all entities connected with the device
+        entity_registry = er.async_get(hass)
+        for platform, entityid in discovery_info.entities:
+            entity_id = entity_registry.async_get_entity_id(platform, DOMAIN, entityid)
+            if entity_id:
+                entity_registry.async_remove(entity_id)
+
+        # destroy the device
+        device_registry = dr.async_get(hass)
+        device_entry = device_registry.async_get_device(
+            identifiers={(DOMAIN, device_id)}
+        )
+        if device_entry:
+            device_registry.async_remove_device(device_entry.id)
+
+        # clean the discovery info
+        del self._discovered[device_id]
+
+    async def start(
+        self, hass: HomeAssistant, mqtt: PyPGLabMqttClient, entry: PGLABConfigEntry
+    ) -> None:
+        """Start to discovery a device."""
+
+        async def discovery_message_received(msg: ReceiveMessage) -> None:
+            """Received a new discovery message."""
+
+            # create a PG LAB device and add in HA all the available entities
+            pglab_device = await self.__build_device(mqtt, msg)
+
+            if not pglab_device:
+                # for some reason it's not possible to create the device with the discovery message
+                # be sure that any previous device with the same topic are now destroy
+                device_id = get_device_id_from_discovery_topic(msg.topic)
+
+                # if we have a valid topic device_id clean every thing relative to the device
+                if device_id:
+                    self.__clean_discovered_device(hass, device_id)
+
+                return
+
+            # we can create a new device now
+            device_registry = dr.async_get(hass)
+            device_registry.async_get_or_create(
+                config_entry_id=entry.entry_id,
+                configuration_url=f"http://{pglab_device.ip}/",
+                connections={(CONNECTION_NETWORK_MAC, pglab_device.mac)},
+                identifiers={(DOMAIN, pglab_device.id)},
+                manufacturer=pglab_device.manufactor,
+                model=pglab_device.type,
+                name=pglab_device.name,
+                sw_version=pglab_device.firmware_version,
+                hw_version=pglab_device.hardware_version,
+            )
+
+            # do some checking if previous entities must be updated
+            if pglab_device.id in self._discovered:
+                # the device is already been discover
+                # get the old discovery info data
+                discovery_info = self._discovered[pglab_device.id]
+
+                if discovery_info.hash == pglab_device.hash:
+                    # best case!!! there is nothing to do ... the device
```

Comment:
remove !!! and ... and replace with proper English

Reply:
Ok